### PR TITLE
POC: EE-2493: mtls for edge agent

### PR DESCRIPTION
**Pull request by SvenDowideit**
# Replaced by https://github.com/portainer/agent/pull/249

again, based on #228 and #229

and requires the matching portainer PR

### CODE REVIEW

File: agent.go [hunk 1]
```diff
@@ -70,6 +70,9 @@ type (
 		EdgeInactivityTimeout string
 		EdgeInsecurePoll      bool
 		LogLevel              string
+		SSLCert               string
+		SSLKey                string
+		SSLCacert             string
```
Comment (cheloRydel) [inline agent.go]:
```suggestion
		SSLCACert             string
```

File: cmd/agent/main.go [hunk 2]
```diff
@@ -3,6 +3,7 @@ package main
 import (
 	"errors"
 	"fmt"
+	http2 "github.com/portainer/agent/edge/http"
```
Comment (cheloRydel) [inline cmd/agent/main.go]:
```suggestion
	httpEdge "github.com/portainer/agent/edge/http"
```

File: edge/edge.go [hunk 3]
```diff
@@ -60,26 +66,44 @@ func (manager *Manager) Start() error {
...
-	log.Printf("[DEBUG] [internal,edge] [api_addr: %s] [edge_id: %s] [poll_frequency: %s] [inactivity_timeout: %s] [insecure_poll: %t]", pollServiceConfig.APIServerAddr, pollServiceConfig.EdgeID, pollServiceConfig.PollFrequency, pollServiceConfig.InactivityTimeout, pollServiceConfig.InsecurePoll)
+	log.Printf("[DEBUG] [internal,edge] [api_addr: %s] [edge_id: %s] [poll_frequency: %s] [inactivity_timeout: %s] [insecure_poll: %t]", pollServiceConfig.APIServerAddr, pollServiceConfig.EdgeID, pollServiceConfig.PollFrequency, pollServiceConfig.InactivityTimeout, manager.agentOptions.EdgeInsecurePoll)
 
-	stackManager, err := newStackManager(manager.key.PortainerInstanceURL, manager.key.EndpointID, manager.agentOptions.EdgeID, pollServiceConfig.InsecurePoll)
+	stackManager, err := stack.NewStackManager(
+		client.NewPortainerClient(
+			manager.key.PortainerInstanceURL,
+			manager.key.EndpointID,
+			manager.agentOptions.EdgeID,
+			GetNewHttpClient(10, manager.agentOptions),
```
Comment (cheloRydel) [inline edge/edge.go:82]: maybe it's cleaner to pass the `manager.agentOptions` as param and call `GetNewHttpClient` internally

File: edge/edge.go [hunk 4]
```diff
@@ -217,11 +236,54 @@ func (manager *Manager) checkDockerRuntimeConfig() error {
...
+		err = manager.stackManager.Stop()
 		if err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
+
+func GetNewHttpClient(timeout float64, options *agent.Options) *http.Client {
```
Comment (cheloRydel) [inline edge/edge.go:247]:
```suggestion
func buildHTTPClient(timeout float64, options *agent.Options) *http.Client {
```

File: edge/edge.go [hunk 5]
```diff
@@ -217,11 +236,54 @@ func (manager *Manager) checkDockerRuntimeConfig() error {
...
+	httpCli := &http.Client{
+		Timeout: time.Duration(timeout) * time.Second,
+	}
+
+	if options.EdgeInsecurePoll {
```
Comment (cheloRydel) [inline edge/edge.go:252]: move all below to an internal function
```suggestion
	httpCli.Transport = buildTransport(options)
	return httpCli
	....
	....
	func buildTransport(options *agent.Options) *http.Transport {
	     if options.SSLCert != "" && options.SSLKey != "" {
	         // return SSL Transport
	     }
	     
	     if options.EdgeInsecurePoll {
	        // return InsecureSkipVerify Transport
	     }
	     
	     return nil
	 }
```

File: edge/edge.go (hunk 3 above)
Comment (SvenDowideit) [reply edge/edge.go:82]: atm, there are too many things created internally, breaking both the sharing, and hiding the high number of internal members - so I've been pulling it appart and making it more obvious.
Annoyingly, what I was originally trying to acheive started to be too big a refactor, so I'm currently doing the async thing in at a PortainerClient level, so its possibly true.
Main thrust is that the actual http requests are now in the pluggable PortainerClient, and this set of 3 managers all use the same code - except for the different timeouts, they would use the same PortainerClient instance - but that's something where I don't have time to work out the consequenses of the two 10second constants vs the User tuned one.

Comment (cheloRydel) [reply edge/edge.go:82]: yep I get that, what I had in mind was to at least pass the `timeout` value as a parameter and use that to call `GetNewHttpClient` internally, but it surely can be changed later, it's mostly a readability issue

File: os/options.go [hunk 6]
```diff
@@ -47,6 +47,13 @@ var (
 	fEdgeInactivityTimeout = kingpin.Flag("edge-inactivity", EnvKeyEdgeInactivityTimeout+" timeout used by the agent to close the reverse tunnel after inactivity (default to 5m)").Envar(EnvKeyEdgeInactivityTimeout).Default(agent.DefaultEdgeSleepInterval).String()
 	fEdgeInsecurePoll      = kingpin.Flag("edge-insecurepoll", EnvKeyEdgeInsecurePoll+" enable this option if you need the agent to poll a HTTPS Portainer instance with self-signed certificates. Disabled by default, set to 1 to enable it").Envar(EnvKeyEdgeInsecurePoll).Bool()
 	fEdgeTunnel            = kingpin.Flag("edge-tunnel", EnvKeyEdgeTunnel+" disable this option if you wish to prevent the agent from opening tunnels over websockets").Envar(EnvKeyEdgeTunnel).Default("true").Bool()
+	
+	// mTLS edge agent certs
+	// TODO: REVIEW
+	// Should t have an env var equivalent?
```
Comment (deviantony) [inline os/options.go]: IMO we should keep consistency by adding env var parameters too.

**Comment by cheloRydel** [comment]
Replaced by https://github.com/portainer/agent/pull/234

**Pull request closed without merging**